File: packages/go-sdk/authoring/steps_test.go
package authoring

import "testing"

func TestJobStep(t *testing.T) {
	s := Job("validate", "job_validate")
	if s.StepRef() != "validate" {
		t.Errorf("expected ref 'validate', got %q", s.StepRef())
	}
	if s.Type() != StepTypeJob {
		t.Errorf("expected type 'job', got %q", s.Type())
	}
	if s.JobID != "job_validate" {
		t.Errorf("expected jobID 'job_validate', got %q", s.JobID)
	}
}

func TestJobStep_WithDependsOn(t *testing.T) {
	s := Job("charge", "job_charge", DependsOn("validate"))
	deps := s.BaseOptions().DependsOn
	if len(deps) != 1 || deps[0] != "validate" {
		t.Errorf("expected dependsOn [validate], got %v", deps)
	}
}

func TestApprovalStep(t *testing.T) {
	timeout := 3600
	s := Approval("review", func(a *ApprovalStep) {
		a.ApprovalTimeoutSecs = &timeout
		a.Approvers = []string{"admin"}
		a.DependsOn = []string{"charge"}
	})

	if s.Type() != StepTypeApproval {
		t.Error("expected approval type")
	}
	if *s.ApprovalTimeoutSecs != 3600 {
		t.Error("expected timeout 3600")
	}
	if len(s.Approvers) != 1 || s.Approvers[0] != "admin" {
		t.Error("expected approvers [admin]")
	}
}

func TestSubWorkflowStep(t *testing.T) {
	depth := 3
	s := SubWorkflow("notify", "wf_notifications", func(sw *SubWorkflowStep) {
		sw.MaxNestingDepth = &depth
	})

	if s.Type() != StepTypeSubWorkflow {
		t.Error("expected sub_workflow type")
	}
	if s.SubWorkflowID != "wf_notifications" {
		t.Error("expected sub workflow ID")
	}
	if *s.MaxNestingDepth != 3 {
		t.Error("expected max nesting depth 3")
	}
}

func TestWaitForEventStep(t *testing.T) {
	timeout := 86400
	s := WaitForEvent("confirm", "shipping.confirmed", func(w *WaitForEventStep) {
		w.EventTimeoutSecs = &timeout
		w.EventNotifyURL = "https://hooks.example.com"
	})

	if s.Type() != StepTypeWaitForEvent {
		t.Error("expected wait_for_event type")
	}
	if s.EventKey != "shipping.confirmed" {
		t.Error("expected event key")
	}
}

func TestSleepStep(t *testing.T) {
	s := Sleep("cooldown", 60)
	if s.Type() != StepTypeSleep {
		t.Error("expected sleep type")
	}
	if s.SleepDurationSecs != 60 {
		t.Error("expected 60 seconds")
	}
}

func TestStepToAPI_Job(t *testing.T) {
	s := Job("validate", "job_validate", DependsOn("init"))
	api := StepToAPI(s)

	if api["step_ref"] != "validate" {
		t.Error("expected step_ref")
	}
	if api["type"] != "job" {
		t.Error("expected type job")
	}
	if api["job_id"] != "job_validate" {
		t.Error("expected job_id")
	}
	deps, ok := api["depends_on"].([]string)
	if !ok || len(deps) != 1 || deps[0] != "init" {
		t.Error("expected depends_on [init]")
	}
}

func TestStepToAPI_Approval(t *testing.T) {
	timeout := 3600
	s := Approval("review", func(a *ApprovalStep) {
		a.ApprovalTimeoutSecs = &timeout
		a.Approvers = []string{"admin"}
	})
	api := StepToAPI(s)

	if api["approval_timeout_secs"] != 3600 {
		t.Error("expected approval_timeout_secs")
	}
	approvers, ok := api["approvers"].([]string)
	if !ok || len(approvers) != 1 {
		t.Error("expected approvers")
	}
}

func TestStepToAPI_SubWorkflow(t *testing.T) {
	depth := 2
	s := SubWorkflow("sub", "wf_1", func(sw *SubWorkflowStep) {
		sw.MaxNestingDepth = &depth
	})
	api := StepToAPI(s)

	if api["sub_workflow_id"] != "wf_1" {
		t.Error("expected sub_workflow_id")
	}
	if api["max_nesting_depth"] != 2 {
		t.Error("expected max_nesting_depth")
	}
}

func TestStepToAPI_WaitForEvent(t *testing.T) {
	s := WaitForEvent("wait", "event.key")
	api := StepToAPI(s)

	if api["event_key"] != "event.key" {
		t.Error("expected event_key")
	}
}

func TestStepToAPI_Sleep(t *testing.T) {
	s := Sleep("pause", 120)
	api := StepToAPI(s)

	if api["sleep_duration_secs"] != 120 {
		t.Error("expected sleep_duration_secs")
	}
}

func TestStepToAPI_BaseOptions(t *testing.T) {
	retryAttempts := 3
	retryDelay := 5
	retryMax := 60
	timeout := 300

	s := Job("process", "job_process", func(o *BaseStepOptions) {
		o.DependsOn = []string{"validate"}
		o.OnFailure = OnFailureFailWorkflow
		o.RetryMaxAttempts = &retryAttempts
		o.RetryBackoff = RetryBackoffExponential
		o.RetryInitialDelaySecs = &retryDelay
		o.RetryMaxDelaySecs = &retryMax
		o.TimeoutSecsOverride = &timeout
		o.OutputTransform = "$.result"
		o.ConcurrencyKey = "order_id"
		o.ResourceClass = ResourceClassLarge
	})

	api := StepToAPI(s)

	if api["on_failure"] != "fail_workflow" {
		t.Error("expected on_failure")
	}
	if api["retry_max_attempts"] != 3 {
		t.Error("expected retry_max_attempts")
	}
	if api["retry_backoff"] != "exponential" {
		t.Error("expected retry_backoff")
	}
	if api["retry_initial_delay_secs"] != 5 {
		t.Error("expected retry_initial_delay_secs")
	}
	if api["retry_max_delay_secs"] != 60 {
		t.Error("expected retry_max_delay_secs")
	}
	if api["timeout_secs_override"] != 300 {
		t.Error("expected timeout_secs_override")
	}
	if api["output_transform"] != "$.result" {
		t.Error("expected output_transform")
	}
	if api["concurrency_key"] != "order_id" {
		t.Error("expected concurrency_key")
	}
	if api["resource_class"] != "large" {
		t.Error("expected resource_class")
	}
}

File: packages/typescript-sdk/test/steps.test.ts
import { describe, expect, test } from "bun:test";

import { step, stepToApi } from "../src/authoring/steps";

describe("step builder", () => {
  test("step.job creates a job step with correct type and fields", () => {
    const s = step.job("validate", "job_validate");
    expect(s.type).toBe("job");
    expect(s.stepRef).toBe("validate");
    expect(s.jobId).toBe("job_validate");
  });

  test("step.job with all base options", () => {
    const s = step.job("charge", "job_charge", {
      dependsOn: ["validate"],
      onFailure: "fail_workflow",
      retryMaxAttempts: 3,
      retryBackoff: "exponential",
      retryInitialDelaySecs: 1,
      retryMaxDelaySecs: 60,
      timeoutSecsOverride: 120,
      outputTransform: "$.result",
      concurrencyKey: "order-{{ orderId }}",
      resourceClass: "large",
      condition: { status: "approved" },
      payload: { key: "value" },
    });

    expect(s.dependsOn).toEqual(["validate"]);
    expect(s.onFailure).toBe("fail_workflow");
    expect(s.retryMaxAttempts).toBe(3);
    expect(s.retryBackoff).toBe("exponential");
    expect(s.retryInitialDelaySecs).toBe(1);
    expect(s.retryMaxDelaySecs).toBe(60);
    expect(s.timeoutSecsOverride).toBe(120);
    expect(s.outputTransform).toBe("$.result");
    expect(s.concurrencyKey).toBe("order-{{ orderId }}");
    expect(s.resourceClass).toBe("large");
    expect(s.condition).toEqual({ status: "approved" });
    expect(s.payload).toEqual({ key: "value" });
  });

  test("step.approval creates an approval step", () => {
    const s = step.approval("review", {
      dependsOn: ["charge"],
      approvalTimeoutSecs: 3600,
      approvers: ["admin@example.com", "manager@example.com"],
    });

    expect(s.type).toBe("approval");
    expect(s.stepRef).toBe("review");
    expect(s.approvalTimeoutSecs).toBe(3600);
    expect(s.approvers).toEqual(["admin@example.com", "manager@example.com"]);
    expect(s.dependsOn).toEqual(["charge"]);
  });

  test("step.subWorkflow creates a sub-workflow step", () => {
    const s = step.subWorkflow("notify-all", "wf_notifications", {
      dependsOn: ["cooldown"],
      maxNestingDepth: 2,
    });

    expect(s.type).toBe("sub_workflow");
    expect(s.stepRef).toBe("notify-all");
    expect(s.subWorkflowId).toBe("wf_notifications");
    expect(s.maxNestingDepth).toBe(2);
  });

  test("step.waitForEvent creates a wait-for-event step", () => {
    const s = step.waitForEvent("shipping", "shipping.confirmed", {
      dependsOn: ["review"],
      eventTimeoutSecs: 86_400,
      eventNotifyUrl: "https://notify.example.com",
    });

    expect(s.type).toBe("wait_for_event");
    expect(s.stepRef).toBe("shipping");
    expect(s.eventKey).toBe("shipping.confirmed");
    expect(s.eventTimeoutSecs).toBe(86_400);
    expect(s.eventNotifyUrl).toBe("https://notify.example.com");
  });

  test("step.sleep creates a sleep step", () => {
    const s = step.sleep("cooldown", 60, { dependsOn: ["shipping"] });

    expect(s.type).toBe("sleep");
    expect(s.stepRef).toBe("cooldown");
    expect(s.sleepDurationSecs).toBe(60);
    expect(s.dependsOn).toEqual(["shipping"]);
  });
});

describe("stepToApi", () => {
  test("converts job step to snake_case API format", () => {
    const api = stepToApi(
      step.job("charge", "job_charge", {
        dependsOn: ["validate"],
        onFailure: "fail_workflow",
        retryMaxAttempts: 3,
        retryBackoff: "exponential",
      })
    );

    expect(api).toEqual({
      step_ref: "charge",
      type: "job",
      job_id: "job_charge",
      depends_on: ["validate"],
      on_failure: "fail_workflow",
      retry_max_attempts: 3,
      retry_backoff: "exponential",
    });
  });

  test("converts approval step to API format", () => {
    const api = stepToApi(
      step.approval("review", {
        approvalTimeoutSecs: 3600,
        approvers: ["admin@example.com"],
      })
    );

    expect(api).toEqual({
      step_ref: "review",
      type: "approval",
      approval_timeout_secs: 3600,
      approvers: ["admin@example.com"],
    });
  });

  test("converts sub-workflow step to API format", () => {
    const api = stepToApi(
      step.subWorkflow("notify", "wf_notify", { maxNestingDepth: 2 })
    );

    expect(api).toEqual({
      step_ref: "notify",
      type: "sub_workflow",
      sub_workflow_id: "wf_notify",
      max_nesting_depth: 2,
    });
  });

  test("converts wait-for-event step to API format", () => {
    const api = stepToApi(
      step.waitForEvent("wait", "order.shipped", {
        eventTimeoutSecs: 86_400,
        eventNotifyUrl: "https://notify.example.com",
      })
    );

    expect(api).toEqual({
      step_ref: "wait",
      type: "wait_for_event",
      event_key: "order.shipped",
      event_timeout_secs: 86_400,
      event_notify_url: "https://notify.example.com",
    });
  });

  test("converts sleep step to API format", () => {
    const api = stepToApi(step.sleep("pause", 30));

    expect(api).toEqual({
      step_ref: "pause",
      type: "sleep",
      sleep_duration_secs: 30,
    });
  });

  test("includes all base step options in API format", () => {
    const api = stepToApi(
      step.job("full", "job_full", {
        retryInitialDelaySecs: 1,
        retryMaxDelaySecs: 60,
        timeoutSecsOverride: 120,
        outputTransform: "$.data",
        concurrencyKey: "key-1",
        resourceClass: "medium",
      })
    );

    expect(api.retry_initial_delay_secs).toBe(1);
    expect(api.retry_max_delay_secs).toBe(60);
    expect(api.timeout_secs_override).toBe(120);
    expect(api.output_transform).toBe("$.data");
    expect(api.concurrency_key).toBe("key-1");
    expect(api.resource_class).toBe("medium");
  });

  test("minimal step only includes required fields", () => {
    const api = stepToApi(step.job("simple", "job_simple"));

    expect(api).toEqual({
      step_ref: "simple",
      type: "job",
      job_id: "job_simple",
    });
  });
});
